internal/runtimehandlerhooks/high_performance_hooks.go

```diff
@@ -356,6 +417,176 @@ func setCPUQuota(cpuMountPoint, parentDir string, c *oci.Container, enable bool)
+			content, err := ioutil.ReadFile(cpuPMQOSResumeLatencyFile)
+			if err != nil {
+				return err
+			}
+			latencyOrig := strings.TrimSpace(string(content))
+
+			// save the current latency so we can restore it later
+			cpuPowerSaveDir := fmt.Sprintf("%s/cpu%d/power", cpuSaveDir, cpu)
+			err = os.MkdirAll(cpuPowerSaveDir, 0o750)
+			if err != nil {
+				return err
+			}
+			cpuLatencyFileOrig := fmt.Sprintf("%s/pm_qos_resume_latency_us", cpuPowerSaveDir)
+			err = ioutil.WriteFile(cpuLatencyFileOrig, []byte(latencyOrig), 0o644)
+			if err != nil {
+				return err
+			}
+
+			// update the pm_qos_resume_latency_us
+			err = ioutil.WriteFile(cpuPMQOSResumeLatencyFile, []byte(latency), 0o644)
+			if err != nil {
+				return err
+			}
+		} else {
+			// retrieve the original latency
+			cpuPMQOSResumeLatencyFileOrig := fmt.Sprintf("%s/cpu%d/power/pm_qos_resume_latency_us", cpuSaveDir, cpu)
+			content, err := ioutil.ReadFile(cpuPMQOSResumeLatencyFileOrig)
+			if err != nil {
+				if errors.Is(err, os.ErrNotExist) {
+					// the latency may have already been restored by a previous invocation of the hook
+					return nil
+				}
+				return err
+			}
+			latencyOrig := strings.TrimSpace(string(content))
+
+			// restore the original latency
+			cpuPMQOSResumeLatencyFile := fmt.Sprintf("%s/cpu%d/power/pm_qos_resume_latency_us", cpuDir, cpu)
+			err = ioutil.WriteFile(cpuPMQOSResumeLatencyFile, []byte(latencyOrig), 0o644)
+			if err != nil {
+				return err
+			}
+
+			// remove the saved latency
+			err = os.Remove(cpuPMQOSResumeLatencyFileOrig)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// setCPUFreqGovernor sets the scaling_governor for a cpu and stores the original
+// value so it can be restored later. If the governor is an empty string, the original
+// scaling_governor value is restored.
+func setCPUFreqGovernor(c *oci.Container, governor, cpuDir, cpuSaveDir string) error {
+	lspec := c.Spec().Linux
+	if lspec == nil ||
+		lspec.Resources == nil ||
+		lspec.Resources.CPU == nil ||
+		lspec.Resources.CPU.Cpus == "" {
+		return errors.Errorf("find container %s CPUs", c.ID())
+	}
+
+	cpus, err := cpuset.Parse(lspec.Resources.CPU.Cpus)
+	if err != nil {
+		return err
+	}
+
+	for _, cpu := range cpus.ToSlice() {
+		if governor != "" {
+			// retrieve the current scaling governor
+			cpuScalingGovernorFile := fmt.Sprintf("%s/cpu%d/cpufreq/scaling_governor", cpuDir, cpu)
+			content, err := ioutil.ReadFile(cpuScalingGovernorFile)
+			if err != nil {
+				return err
+			}
+			governorOrig := strings.TrimSpace(string(content))
+
+			// get available cpu scaling governors
+			cpuScalingAvailableGovernorFile := fmt.Sprintf("%s/cpu%d/cpufreq/scaling_available_governors", cpuDir, cpu)
+			content, err = ioutil.ReadFile(cpuScalingAvailableGovernorFile)
+			if err != nil {
+				return err
+			}
+			availableGovernors := strings.Split(strings.TrimSpace(string(content)), " ")
```

Comment:
Now, all CPUs have the same available governors, so you don't have to read it for every CPUI -- can only read it once.

Reply:
I addressed the first two items (using strings.Fields() and removing the intermediate varaible). However, although all CPUs currently have the same available governors, I'd rather not make that assumption in the code, since sysfs allows these to be different for each CPU.